luigi/contrib/snowflake.py

```diff
@@ -0,0 +1,523 @@
+        """
+        Override to return true if table should be truncated before copying new
+        data in.
+        """
+        return False
+
+    @property
+    def queries(self):
+        """
+        Override to return a list of queries to be executed in order.
+        """
+        return []
+
+    def truncate_table(self, connection):
+        """
+        Truncate the table
+        """
+        query = "truncate {table}".format(table=self.table)
+        cursor = connection.cursor()
+        try:
+            cursor.execute(query)
+        finally:
+            cursor.close()
+
+    def create_schema(self, connection):
+        """
+        Will create the schema in the database if it does not exists.
+
+        If no schema is detected in the table string, then we assume that it
+        uses the default schema of the database which is "public" if not
+        overriden.
+
+        The default schema should always exists, so there is no point in
+        creating it.
+        """
+        if '.' not in self.table:
+            return
+
+        query = 'CREATE SCHEMA IF NOT EXISTS {schema_name};'.format(schema_name=self.table.split('.')[0])
+        connection.cursor().execute(query)
+
+    def run(self):
+        """
+        COPY data residing on S3 to a Snowflake table following these steps:
+
+        * Fetch the file path from S3
+        * Define the ouput as a Snowflake Target
+        * Create a connection and a cursor to copy data to Snowflake
+
+        * Create the schema if it doesn't already exists
+        * Create the table if it doesn't already exists
+        * Truncate the table if the flag "do_truncate_table" is set
+
+        * Copy the data from S3 to the Snowflake table
+
+        * Run any potential "post_copy" queries
+
+        * Save the state of this task to Snowflake as "been run"
+        """
+        path = self.s3_load_path()
+        output = self.output()
+        connection = output.connect()
+        cursor = connection.cursor()
+
+        self.init_copy(connection)
+        self.copy(cursor, path)
+        self.post_copy(cursor)
+
+        # update marker table
+        output.touch()
+        connection.commit()
+
+        # commit and clean up
+        connection.close()
+
+    def copy(self, cursor, f):
+        """
+        COPY data from from S3 into Snowflake.
+
+        This currently only works for key-base credentials.
+        """
+        logger.info("Inserting file: {file}".format(file=f))
+        colnames = ''
+        if self.columns and len(self.columns) > 0:
+            colnames = ",".join([x[0] for x in self.columns])
+            colnames = '({})'.format(colnames)
+
+        query = ("""COPY INTO {table} from {source} CREDENTIALS=({creds}) {options};""".format(
+            table=self.table,
+            source=f,
+            creds=self.credentials(),
+            options=self.copy_options)
+        )
+        cursor.execute(query)
+
+    def credentials(self):
```

Comment:
in `redshift.py`, credentials is a hidden method. What do you think is the appropriate type? hidden vs public? method vs property?